### PR TITLE
ENH: allow opt-in to inferring pyarrow strings

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This is a simple opt-in for now without warning by default. Can adjust this for 2.2

cc @jbrockmendel

### CODE REVIEW

File: pandas/tests/frame/test_constructors.py [hunk 1]
```diff
@@ -2670,6 +2670,41 @@ def test_construct_with_strings_and_none(self):
...
+        )
+        with pd.option_context("future.infer_string", True):
+            df = DataFrame({"a": ["a", 1]})
+        tm.assert_frame_equal(df, expected)
+
+        expected = DataFrame(
+            {"a": ["a", "b"]}, dtype="object", columns=Index(["a"], dtype=dtype)
+        )
+        with pd.option_context("future.infer_string", True):
+            df = DataFrame({"a": ["a", "b"]}, dtype="object")
```
Comment (lithomas1) [inline pandas/tests/frame/test_constructors.py:2705]: Can you add a case with null/nan/None in it?

Comment (phofl) [reply pandas/tests/frame/test_constructors.py:2705]: Added, needed a fix...

File: pandas/tests/io/parser/dtypes/test_dtypes_basic.py [hunk 2]
```diff
@@ -538,3 +538,24 @@ def test_ea_int_avoid_overflow(all_parsers):
...
+    # GH#54430
+    pa = pytest.importorskip("pyarrow")
+    dtype = pd.ArrowDtype(pa.string())
+
+    data = """a,b
+x,1
+y,2"""
+    parser = all_parsers
+    if parser.engine == "pyarrow":
+        pytest.skip("TODO: Follow up")
```
Comment (WillAyd) [inline pandas/tests/io/parser/dtypes/test_dtypes_basic.py:553]: Did we ever discuss pyarrow backed engines returning pyarrow types by default? I think from a user perspective it is less than ideal to have to specify both this option and the dtype backend for any pyarrow backed IO methods

**Review by WillAyd — COMMENTED** [review]
Looks pretty good. Have we considered naming the option `infer_pyarrow_string` instead of `infer_string`?

File: pandas/tests/io/parser/dtypes/test_dtypes_basic.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/io/parser/dtypes/test_dtypes_basic.py:553]: Let's move this discussion to Basel, but there is an issue about this: https://github.com/pandas-dev/pandas/issues/51846

**Comment by phofl** [comment]
cc @jbrockmendel thoughts about the option name?

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -796,6 +798,12 @@ def infer_dtype_from_scalar(val) -> tuple[DtypeObj, Any]:
         # coming out as np.str_!
 
         dtype = _dtype_obj
+        opt = get_option("future.infer_string")
+        if opt is True:
+            import pyarrow as pa
+
+            pa_dtype = pa.string()
+            dtype = ArrowDtype(pa_dtype)
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:806]: are there any tests that hit this?

Comment (phofl) [reply pandas/core/dtypes/cast.py:806]: Added one

**Review by jbrockmendel — APPROVED** [review]

**Review by mroeschke — COMMENTED** [review]
Could use some documentation in the whatsnew otherwise looks good

**Comment by phofl** [comment]
I’ll add this With the general arrow announcement

**Review by mroeschke — APPROVED** [review]

**Pull request merged**